src/rewrite_reduce.cpp

```diff
@@ -0,0 +1,92 @@
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.  IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ *
+ */
+#include <migraphx/rewrite_reduce.hpp>
+#include <migraphx/module.hpp>
+#include <migraphx/matcher.hpp>
+#include <migraphx/make_op.hpp>
+
+namespace migraphx {
+inline namespace MIGRAPHX_INLINE_NS {
+
+namespace {
+struct find_softmax
+{
+    auto matcher() const { return match::name("softmax"); }
+
+    void apply(module& m, const match::matcher_result& r) const
+    {
+        auto ins  = r.result;
+        auto op   = ins->get_operator().to_value();
+        auto axis = op["axis"].to<std::int64_t>();
+
+        auto input = ins->inputs().front();
+        auto max   = m.insert_instruction(ins, make_op("reduce_max", {{"axes", {axis}}}), input);
+        auto maxb  = m.insert_instruction(
+            ins, make_op("multibroadcast", {{"out_lens", input->get_shape().lens()}}), max);
+        auto sub  = m.insert_instruction(ins, make_op("sub"), input, maxb);
+        auto exp  = m.insert_instruction(ins, make_op("exp"), sub);
+        auto sum  = m.insert_instruction(ins, make_op("reduce_sum", {{"axes", {axis}}}), exp);
+        auto sumb = m.insert_instruction(
+            ins, make_op("multibroadcast", {{"out_lens", input->get_shape().lens()}}), sum);
+        m.replace_instruction(ins, make_op("div"), exp, sumb);
+    }
+};
+
+struct rewrite_reduce_mean
+{
+    auto matcher() const { return match::name("reduce_mean"); }
+
+    void apply(module& m, const match::matcher_result& r) const
+    {
+        auto reduce_mean = r.result;
+        auto op          = reduce_mean->get_operator().to_value();
+        auto axes        = op["axes"].to_vector<std::int64_t>();
+        auto input       = reduce_mean->inputs().front();
+
+        auto dims = input->get_shape().lens();
+        double n  = std::accumulate(
+            dims.begin() + 2, dims.end(), 1, [&](const auto& i, const auto& j) { return i * j; });
+        auto n_literal = m.add_literal(literal{{input->get_shape().type(), {1}}, {1.0 / n}});
```

Comment:
This wont work for integers. This should be `n` and not `1/n` to make it simpler for the integer path(simplify_algebra can rewrite to the reciprocal).